src/server/coprocessor.rs

```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
+    }
+    Ok(rows)
+}
+
+fn prefix_next(key: &[u8]) -> Vec<u8> {
+    let mut nk = key.to_vec();
+    if nk.is_empty() {
+        nk.push(0);
+        return nk;
+    }
+    let mut i = nk.len() - 1;
+    loop {
+        if nk[i] == 255 {
+            nk[i] = 0;
+        } else {
+            nk[i] += 1;
+            return nk;
+        }
+        if i == 0 {
+            nk = key.to_vec();
+            nk.push(0);
+            return nk;
+        }
+        i -= 1;
+    }
+}
+
+/// `is_point` checks if the key range represents a point.
+fn is_point(range: &KeyRange) -> bool {
+    range.get_end() == &*prefix_next(range.get_start())
+}
+
+fn get_rows_from_range(store: &SnapshotStore,
+                       sel: &SelectRequest,
+                       mut range: KeyRange,
+                       eval: &mut Evaluator)
+                       -> Result<Vec<Row>> {
+    let mut rows = vec![];
+    if is_point(&range) {
+        if let None = try!(store.get(&Key::from_raw(range.get_start().to_vec()))) {
+            return Ok(rows);
+        }
+        let h = box_try!(table::decode_handle(range.get_start()));
+        if let Some(row) = try!(get_row_by_handle(store, sel, h, eval)) {
+            rows.push(row);
+        }
+    } else {
+        let mut seek_key = range.take_start();
+        loop {
+            let mut res = try!(store.scan(Key::from_raw(seek_key), 1));
+            if res.is_empty() {
+                break;
+            }
+            let (key, _) = try!(res.pop().unwrap());
+            if range.get_end() < &key {
+                break;
+            }
+            let h = box_try!(table::decode_handle(&key));
+            if let Some(row) = try!(get_row_by_handle(store, sel, h, eval)) {
+                rows.push(row);
+            }
+            seek_key = prefix_next(&key);
+        }
+    }
+    Ok(rows)
+}
+
+fn get_row_by_handle(store: &SnapshotStore,
+                     sel: &SelectRequest,
+                     h: i64,
+                     eval: &mut Evaluator)
+                     -> Result<Option<Row>> {
+    let tid = sel.get_table_info().get_table_id();
+    let columns = sel.get_table_info().get_columns();
+    let mut row = Row::new();
+    let handle = box_try!(datum::encode_value(&[Datum::I64(h)]));
+    for col in columns {
+        if col.get_pk_handle() {
+            row.mut_data().extend(handle.clone());
+        } else {
+            let raw_key = table::encode_column_key(tid, h, col.get_column_id());
+            let key = Key::from_raw(raw_key);
+            match try!(store.get(&key)) {
+                None => return Err(box_err!("key {:?} not exists", key)),
+                Some(bs) => row.mut_data().extend(bs),
+            }
+        }
+    }
+    row.set_handle(handle);
+    if !sel.has_field_where() {
+        return Ok(Some(row));
+    }
+    if !row.get_data().is_empty() {
+        let (datums, _) = box_try!(datum::decode(row.get_data()));
+        for (c, d) in columns.iter().zip(datums) {
```

Comment:
Is `zip` for data compression here? What is the time cost for `zip` function call?